<a href="https://colab.research.google.com/github/gorand850/machine-learning-bootcamp/blob/main/06_Regresja_liniowa/32_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


#scikit-learn

Strona biblioteki: https://scikit-learn.org

Dokumentacja/User Guide: https://scikit-learn.org/stable/user_guide.html

Podstawowa biblioteka do uczenia maszynowego w języku Python.

Aby zainstalować bibliotekę scikit-learn, użyj polecenia poniżej:

`!pip install scikit-learn`

Aby zaktualizować do najnowszej wersji bibliotekę scikit-learn, użyj polecenia poniżej:

`!pip install --upgrade scikit-learn`

Kurs stworzony w oparciu o wersję 0.22.1
Spis treści:

1. [Import bibliotek](#0)
2. [Przygotowanie danych](#1)
3. [Równanie normalne](#2)
4. [Regresja liniowa przy pomocy scikit-learn](#3)



###<a name="0"></a>Import bibliotek

In [1]:
import numpy as np

###<a name="1"></a>Przygotowanie danych

In [2]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [3]:
#Zmiana formatu danych, z tablicy 1 wymoiarowej, na tablice 6 wymiarową z 1 elementem
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [4]:
#generujemy wektor wyrazów wolnych, które przydadzą nam się do wyliczania stałej w równaniu, często nazywany bias (bajas)
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [5]:
#łączymy wektory X1 i bias, tak by była to macierz zmiennych
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


###<a name="2"></a>Równanie normalne
Regresja liniowa w $R^2$: 
$$Y = w_0 + w_1X_1$$ 
$$Y = W^TX$$ 
gdzie: 
$$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ 
stąd $$ W^T= \left[\begin{matrix}w_0&amp;w_1\end{matrix}\right] $$
$$Y = W^TX = \left[\begin{matrix}w_0&amp;w_1\end{matrix}\right] \cdot \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje) 
$$ W = (X^TX)^{-1} (X^TY) $$

In [6]:
#mnożenie macieży w numpy za pomocą metody dot
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

obliczamy wartość równania: $(X^TX)^{-1}$

In [7]:
L = np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

obliczamy wartość równania: $(X^TY)$

In [8]:
P = np.dot(X.T,Y)
P

array([21750., 80500.])

Mnożymy macierze $W = LP = (X^TX)^{-1}(X^TY) $

In [9]:
np.dot(L, P)

array([2750.,  250.])

Końcowa postać modelu: $Y = 2750 + 250X_1$

##<a name="3"></a>Regresja liniowa przy pomocy scikit-learn

In [10]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
